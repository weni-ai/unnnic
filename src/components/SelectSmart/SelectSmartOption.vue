<template>
  <div
    :class="{
      'unnnic-select-smart-option': true,
      'unnnic-select-smart-option--focused': focused,
      'unnnic-select-smart-option--active': active,
      'unnnic--clickable': selectable,
      'unnnic-select-smart-option--selectable': selectable,
      'unnnic-select-smart-option--with-checkbox': allowCheckbox,
    }"
    :title="label"
    data-testid="select-smart-option"
  >
    <UnnnicCheckbox
      v-if="allowCheckbox"
      ref="checkbox"
      data-testid="checkbox"
      :modelValue="active"
      :size="size"
    />
    <div>
      <span
        :class="[
          'unnnic-select-smart-option__label',
          `unnnic-select-smart-option__label--${size}`,
        ]"
        data-testid="label"
        >{{ sanitizedValue(label) }}</span
      >
      <p
        v-if="description"
        data-testid="description"
        :class="[
          'unnnic-select-smart-option__description',
          `unnnic-select-smart-option__description--${size}`,
        ]"
      >
        {{ sanitizedValue(description) }}
      </p>
    </div>
  </div>
</template>

<script>
import UnnnicCheckbox from '../Checkbox/Checkbox.vue';
import { escapeHtml } from '../../utils/sanitize'

export default {
  name: 'UnnnicSelectSmartOption',
  components: { UnnnicCheckbox },
  props: {
    label: {
      type: String,
      default: 'Option',
      required: true,
    },
    description: {
      type: String,
      default: '',
      required: false,
    },
    size: {
      type: String,
      default: '',
    },
    selectable: {
      type: Boolean,
      default: true,
    },
    focused: {
      type: Boolean,
      default: false,
    },
    active: {
      type: Boolean,
      default: null,
    },
    allowCheckbox: {
      type: Boolean,
      default: false,
    },
    isMultiple: Boolean,
  },
  methods: {
    sanitizedValue(value) {
      return escapeHtml(value);
    }
  }
};
</script>

<style lang="scss">
@import '../../assets/scss/unnnic.scss';
.unnnic-select-smart-option {
  background-color: $unnnic-color-background-snow;
  padding: $unnnic-spacing-stack-nano $unnnic-inline-xs;

  max-width: 100%;

  font-family: $unnnic-font-family-secondary;
  color: $unnnic-color-neutral-darkest;
  font-weight: $unnnic-font-weight-regular;

  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  -webkit-line-clamp: 1;

  & + & {
    margin-top: $unnnic-spacing-nano;
  }

  &--selectable:hover,
  &--focused {
    border-radius: $unnnic-border-radius-sm;

    background-color: $unnnic-color-neutral-light;
  }

  &--active:not(&--with-checkbox) {
    color: $unnnic-color-weni-600;
    font-weight: $unnnic-font-weight-bold;
  }

  &--with-checkbox {
    padding: $unnnic-spacing-xs $unnnic-spacing-ant;
    display: flex;
    align-items: center;
    gap: $unnnic-spacing-xs;

    .unnnic-checkbox rect {
      rx: $unnnic-border-radius-md;
    }
  }

  &__label {
    display: block;

    &--md {
      font-size: $unnnic-font-size-body-gt;
      line-height: $unnnic-font-size-body-gt + $unnnic-line-height-medium;
    }
    &--sm {
      font-size: $unnnic-font-size-body-md;
      line-height: $unnnic-font-size-body-md + $unnnic-line-height-md;
    }
  }

  &__description {
    margin: 0;

    color: $unnnic-color-neutral-cloudy;
    font-weight: $unnnic-font-weight-regular;

    &--md,
    &--sm {
      font-size: $unnnic-font-size-body-md;
      line-height: $unnnic-font-size-body-md + $unnnic-line-height-medium;
    }
  }
}
</style>
