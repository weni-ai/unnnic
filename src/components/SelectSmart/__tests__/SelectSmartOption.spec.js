import { shallowMount } from '@vue/test-utils';
import SelectSmartOption from '../SelectSmartOption.vue';

describe('SelectSmartOption.vue', () => {
  let wrapper;

  const mountWrapper = (props = {}) => {
    wrapper = shallowMount(SelectSmartOption, {
      propsData: {
        label: 'Test Option',
        ...props,
      },
    });
  };

  beforeEach(() => {
    mountWrapper();
  });

  describe('Initial rendering', () => {
    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('renders the component with the correct label', () => {
      expect(wrapper.find('[data-testid="label"]').text()).toBe('Test Option');
    });

    it('applies the title attribute with the label text', () => {
      expect(wrapper.attributes('title')).toBe('Test Option');
    });
  });

  describe('Dynamic classes', () => {
    it('adds focused class when focused prop is true', async () => {
      await wrapper.setProps({ focused: true });
      expect(wrapper.classes()).toContain(
        'unnnic-select-smart-option--focused',
      );
    });

    it('adds active class when active prop is true', async () => {
      await wrapper.setProps({ active: true });
      expect(wrapper.classes()).toContain('unnnic-select-smart-option--active');
    });

    it('adds selectable class when selectable prop is true', async () => {
      await wrapper.setProps({ selectable: true });
      expect(wrapper.classes()).toContain(
        'unnnic-select-smart-option--selectable',
      );
    });

    it('adds with-checkbox class when allowCheckbox prop is true', async () => {
      await wrapper.setProps({ allowCheckbox: true });
      expect(wrapper.classes()).toContain(
        'unnnic-select-smart-option--with-checkbox',
      );
    });
  });

  describe('Checkbox rendering', () => {
    beforeEach(() => {
      mountWrapper({ allowCheckbox: true });
    });

    it('renders checkbox when allowCheckbox is true', () => {
      const checkbox = wrapper.findComponent('[data-testid="checkbox"]');
      expect(checkbox.exists()).toBe(true);
    });

    it('does not render checkbox when allowCheckbox is false', () => {
      mountWrapper({ allowCheckbox: false });
      const checkbox = wrapper.findComponent('[data-testid="checkbox"]');
      expect(checkbox.exists()).toBe(false);
    });

    it('binds active prop to checkbox modelValue', async () => {
      await wrapper.setProps({ active: true });
      const checkbox = wrapper.findComponent('[data-testid="checkbox"]');
      expect(checkbox.props('modelValue')).toBe(true);
    });
  });

  describe('Label and description rendering', () => {
    it('displays the label with the correct size class', async () => {
      await wrapper.setProps({ size: 'large' });
      const label = wrapper.find('[data-testid="label"]');
      expect(label.classes()).toContain(
        'unnnic-select-smart-option__label--large',
      );
    });

    it('displays the description if provided', async () => {
      await wrapper.setProps({ description: 'This is a test description' });
      const description = wrapper.find('[data-testid="description"]');
      expect(description.exists()).toBe(true);
      expect(description.text()).toBe('This is a test description');
    });

    it('does not render description if it is not provided', () => {
      const description = wrapper.find('[data-testid="description"]');
      expect(description.exists()).toBe(false);
    });

    it('applies size class to description if provided', async () => {
      await wrapper.setProps({ description: 'Description', size: 'small' });
      const description = wrapper.find('[data-testid="description"]');
      expect(description.classes()).toContain(
        'unnnic-select-smart-option__description--small',
      );
    });
  });
});
