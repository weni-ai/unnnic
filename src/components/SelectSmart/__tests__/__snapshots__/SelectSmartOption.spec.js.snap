// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SelectSmartOption.vue > Initial rendering > matches the snapshot 1`] = `
"<div class="unnnic-select-smart-option unnnic--clickable unnnic-select-smart-option--selectable" title="Test Option" data-testid="select-smart-option">
  <!--v-if-->
  <div><span class="unnnic-select-smart-option__label unnnic-select-smart-option__label--" data-testid="label">Test Option</span>
    <!--v-if-->
  </div>
</div>"
`;
